src/fe/fe.C

```diff
@@ -202,45 +216,63 @@ void FE<Dim,T>::reinit(const Elem * elem,
               this->_elem_type = elem->type();
               this->_elem_p_level = elem->p_level();
               this->_p_level = this->_add_p_level_in_reinit * elem->p_level();
+
               // Initialize the shape functions
               this->_fe_map->template init_reference_to_physical_map<Dim>
                 (this->qrule->get_points(), elem);
               this->init_shape_functions (this->qrule->get_points(), elem);
-
-              if (this->shapes_need_reinit())
-                {
-                  cached_nodes.resize(elem->n_nodes());
-                  for (auto n : elem->node_index_range())
-                    cached_nodes[n] = elem->point(n);
-                }
             }
           else
             {
-              // libmesh_assert_greater (elem->n_nodes(), 1);
               this->_elem = elem;
 
-              cached_nodes_still_fit = true;
-              if (cached_nodes.size() != elem->n_nodes())
-                cached_nodes_still_fit = false;
-              else
-                for (auto n : make_range(1u, elem->n_nodes()))
-                  {
-                    if (!(elem->point(n) - elem->point(0)).relative_fuzzy_equals
-                        ((cached_nodes[n] - cached_nodes[0]), 1e-13))
-                      {
-                        cached_nodes_still_fit = false;
-                        break;
-                      }
-                  }
-
-              if (this->shapes_need_reinit() && !cached_nodes_still_fit)
+              // Check if cached elem geometry is sufficiently similar
+              cached_elem_still_fits = cached_nodes.size() == elem->n_nodes();
+              for (unsigned n = 1; cached_elem_still_fits && n < elem->n_nodes(); n++)
+                cached_elem_still_fits = (elem->point(n) - elem->point(0)).relative_fuzzy_equals
+                                         (cached_nodes[n] - cached_nodes[0]);
+
+              // Check if cached edge and face orientations are equal
+              if constexpr (T == HIERARCHIC || T == L2_HIERARCHIC || T == HIERARCHIC_VEC ||
+                            T == L2_HIERARCHIC_VEC || T == BERNSTEIN || T == SZABAB ||
+                            T == NEDELEC_ONE || T == RAVIART_THOMAS)
```

Comment:
Could we move this logic into a static Elem:: array?

Reply:
>I also think the cache validity checking is getting to a level of complexity that deserves being broken out into its own method

Done.

> Could we move this logic into a static Elem:: array?

Moved to an `FEInterface` member function instead as suggested in (2) above.